qlib/contrib/model/pytorch_gru.py

```diff
@@ -0,0 +1,362 @@
+            self.train_optimizer.zero_grad()
+
+            choice = np.random.choice(train_num, self.batch_size)
+            x_batch_auto = x_train_values[choice]
+            y_batch_auto = y_train_values[choice]
+
+            if self.use_gpu:
+                x_batch_auto = x_batch_auto.float().cuda()
+                y_batch_auto = y_batch_auto.float().cuda()
+
+            # forward
+            preds = self.gru_model(x_batch_auto)
+            cur_loss = self.get_loss(preds, y_batch_auto, self.loss_type)
+            cur_loss.backward()
+            self.train_optimizer.step()
+            loss.update(cur_loss.item())
+
+            # validation
+            train_loss += loss.val
+            # print(loss.val)
+            if step and step % self.eval_steps == 0:
+                stop_steps += 1
+                train_loss /= self.eval_steps
+
+                with torch.no_grad():
+                    self.gru_model.eval()
+                    loss_val = AverageMeter()
+
+                    # forward
+                    preds = self.gru_model(x_val_auto)
+                    cur_loss_val = self.get_loss(preds, y_val_auto, self.loss_type)
+                    loss_val.update(cur_loss_val.item())
+
+                if verbose:
+                    self.logger.info(
+                        "[Epoch {}]: train_loss {:.6f}, valid_loss {:.6f}".format(step, train_loss, loss_val.val)
+                    )
+                evals_result["train"].append(train_loss)
+                evals_result["valid"].append(loss_val.val)
+                if loss_val.val < best_loss:
+                    if verbose:
+                        self.logger.info(
+                            "\tvalid loss update from {:.6f} to {:.6f}, save checkpoint.".format(
+                                best_loss, loss_val.val
+                            )
+                        )
+                    best_loss = loss_val.val
+                    stop_steps = 0
+                    torch.save(self.gru_model.state_dict(), save_path)
+                train_loss = 0
+                # update learning rate
+                self.scheduler.step(cur_loss_val)
+
+        # restore the optimal parameters after training ??
+        # self.gru_model.load_state_dict(torch.load(save_path))
+        if self.use_gpu:
+            torch.cuda.empty_cache()
+
+    def get_loss(self, pred, target, loss_type):
+        if loss_type == "mse":
+            sqr_loss = (pred - target)**2
+            loss = sqr_loss.mean()
+            return loss
+        elif loss_type == "binary":
+            loss = nn.BCELoss()
+            return loss(pred, target)
+        else:
+            raise NotImplementedError("loss {} is not supported!".format(loss_type))
+
+    def predict(self, dataset):
+        if not self._fitted:
+            raise ValueError("model is not fitted yet!")
+
+        x_test = dataset.prepare("test", col_set="feature")
+        x_test = x_test.fillna(0)
```

Comment:
Same as above.
It will be better to implement it in specific dataset or handler